### ISSUE TITLE
Jax is somehow much slower than numpy!

### ISSUE THREAD

**Issue by sherlcok314159** [google/jax#14620]
### Description

Hi, I am new to Jax and attracted by its elegance. I implement `Softmax Regression` and compare the speed on two frameworks: jax & numpy. However, I am surprised to see jax is much slower than numpy. When in numpy, just remove everything related to jax and I do not put it for convenience. When running, there is no error or warning.

For anyone to reproduce this error, you can download data from [here](https://drive.google.com/drive/folders/18RMnSNCQfnnTKCovzM3BMCx6vU75RxaS?usp=sharing).

```python
import time
from jax import jit
import numpy as np
import pandas as pd
import jax.numpy as jnp

class SoftmaxRegression:
    """Implementation of Softmax Regression.
    
    Usage:
        >>> model = SoftmaxRegression(k=10)
        >>> model.fit(x_train, y_train)
        >>> model.predict(x_eval)
    """
    def __init__(self, k=2, lr=0.2, eps=1e-5, theta=None):
        self.k = k
        self.lr = lr
        self.eps = eps
        self.theta = theta

    def fit(self, x, y):
        """Using Gradient Descent to minimize J(theta) for softmax regression.

        Args:
            x: inputs of shape (m, n).
            y: labels of shape (m, k) k = #classes.
        """
        def softmax(x):
            """Implementation of softmax function.
            Args:
                x: inputs of shape (m, k).

            Returns:
                Outputs of shape (m, k).
            """
            exp = jnp.exp(x)
            return exp / exp.sum(axis=1)[:, None]

        def h(theta, x):
            """
            Args:
                x: inputs of shape (m, n).
                theta: parameters of shape (k, n).
            Returns:
                Outputs of shape (m, k).
            """
            return softmax(x @ theta.T)
        
        def gradient(theta, x, y):
            """
            Args:
                x: inputs of shape (m, n).
                theta: parameters of shape (k, n).
                y: labels of shape (m, k).
            Returns:
                Outputs of shape (k, n).
            """
            m, _ = x.shape
            return (h(theta, x) - y).T @ x / m
        
        def next_theta(theta, x, y):
            """The next theta updated by Gradient Descent.

             Args:
                x: inputs of shape (m, n).
                theta: params of shape(k, n). 
                y: labels of shape (m, k).

            Returns:
                outputs of shape (k, n).
            """
            return theta - self.lr * gradient(theta, x, y)
        
        m, n = x.shape
        # initialize the theta
        if self.theta is None:
            self.theta = jnp.zeros((self.k, n))
        
        old_theta = self.theta
        new_theta = next_theta(self.theta, x, y)
        while jnp.linalg.norm(new_theta - old_theta, ord=1) >= self.eps:
            old_theta = new_theta
            new_theta = next_theta(old_theta, x, y)
        
        self.theta = new_theta

    def predict(self, x):
        """Make predictions given new inputs.
        
        Args:
            x: Inputs of shape (m, n).
        
        Returns:
            Outputs of shape (m, ).
        """
        return jnp.argmax(x @ self.theta.T, axis=1)

def one_hot(x, k):
    m = len(x)
    encoded = np.zeros((m, k))
    encoded[range(m), x]= 1
    # encoded.at[range(m), x].set(1)
    return encoded

def normalize(x):
    mean = x.mean()
    std = x.std()
    return (x - mean) / std

train_df = pd.read_csv('./mnist/train.csv')
test_df = pd.read_csv('./mnist/test.csv')

read_label = lambda x: x['label'].to_numpy() 
read_inputs = lambda x: normalize(x.drop(columns=['label']).to_numpy())

y_train, y_test = map(read_label, (train_df, test_df))
x_train, x_test = map(read_inputs, (train_df, test_df))

y_train = one_hot(y_train, k=10)

# too much data can cause MemoryError for numpy runs on CPU
x_train, y_train = x_train[:2000], y_train[:2000]

model = SoftmaxRegression(k=10)
begin = time.time()
model.fit(x_train, y_train)
y_pred = model.predict(x_test)
acc = jnp.sum(y_pred == y_test) / len(y_pred)
print('Acc is %f'%acc)
end = time.time()
print('Numpy Time is %f'%(end-begin))
```
![image](https://user-images.githubusercontent.com/76043326/220564510-4c7a00f7-ae50-4ced-8a14-3b37a138eddf.png)

### What jax/jaxlib version are you using?

jax v0.3.24; jaxlib v0.3.24; tensorflow: 2.4.1

### Which accelerator(s) are you using?

GPU

### Additional system info

Python 3.8.13 on Ubuntu 20.04

### NVIDIA GPU info

The two GPUS are `RTX 2080TI`. And when running, one GPU just took about 10000 MiB and its GPU-util is 24%

![image](https://user-images.githubusercontent.com/76043326/220565014-23d63292-a44a-4b37-b0f2-d307242c99a0.png)

**Comment by apaszke** [comment]
I think this is expected to a large degree. JAX doesn't come with any precompiled code that does the array operations, it has to compile it once you actually start running the code. This is fine for long-running processes that e.g. involve repeated iterations of training loops, because you only pay the cost of compilation once. But, if you only run the program once then this probably is the major source of overheads.

Beyond that, I would recommend you wrap the bulk of your code in `jax.jit` which will compile a number of operations jointly, for better performance.

That's about as many hints as I have for now. The example is not self-contained and I can't easily see what the shapes of intermediates are. There are certainly workloads for which CPUs are faster than GPUs and you might be in one of those regimes too. All in all, I don't think this is a JAX bug, but please reopen if you see anything suspicious.

**Comment by jakevdp** [comment]
Regarding what to expect in general in speed comparisons between JAX and NumPy, this FAQ entry may be helpful: https://jax.readthedocs.io/en/latest/faq.html#is-jax-faster-than-numpy
